background.js

```diff
@@ -1,49 +1,82 @@
+  const retryDelay = 100;
+
+  let timeLeft = 1000;
+  let [isElement] = await browser.tabs.executeScript({code: elementFoundScript});
+
+  while (!isElement && timeLeft > 0) {
+    await new Promise(r => setTimeout(r, retryDelay));
+    [isElement] = await browser.tabs.executeScript({code: elementFoundScript});
+    timeLeft -= retryDelay;
+  }
 
-  await new Promise(r => setTimeout(r, 500));
+  if (!isElement)
+    showMessage(`Cannot find \`${querySelector}\` element!`);
+
+  return isElement;
+}
+
+const waitAndClick = async querySelector => {
+  if (await waitForElement(querySelector))
+    await clickElement(querySelector);
+}
+
+browser.pageAction.onClicked.addListener(async () => {
+  await browser.tabs.executeScript({code: `
+    if (!document.querySelector(".caal-alerts")) {
+      alerts = document.createElement("div")
+      alerts.setAttribute("class", "caal-alerts");
+      alerts.setAttribute("style", "position:absolute;");
+      document.body.appendChild(alerts);
+    }
+  `});
 
-  await browser.tabs.executeScript({
-    code: `document.getElementsByClassName("board-tab-item-underlined-component share-menu-tab-selector-tab")[0].click();`,
-  });
+  await waitAndClick(".download");
 
-  await browser.tabs.executeScript({
-    code: `e = document.querySelector("[name='isComputerAnalysisEnabled']"); if (e !== null) e.click();`,
-  });
+  await waitAndClick(".board-tab-item-underlined-component");
 
-  setTimeout(async () => {
-    const [pgn] = await browser.tabs.executeScript({
-      code: `document.querySelector("[name='pgn']").value;`,
-    });
-    
-    await browser.tabs.executeScript({
-      code: `document.getElementsByClassName("icon-font-chess x ui_outside-close-icon")[0].click();`,
-    });
+  if (await waitForElement("[name='pgn']")) {
+    const [pgn] = await browser.tabs.executeScript({code: `document.querySelector("[name='pgn']").value;`});
 
-    await browser.tabs.create({ url: "https://lichess.org/paste" });
+    await waitAndClick("[data-cy='share-menu-close']");
 
-    await new Promise(r => setTimeout(r, 500));
+    await browser.tabs.create({url: "https://lichess.org/paste"});
 
-    const [logged] = await browser.tabs.executeScript({
-      code: `!document.getElementById("form3-analyse").disabled;`
-    });
+    if (await waitForElement("[name='analyse']")) {
+      const [loggedIn] = await browser.tabs.executeScript({code: `!document.querySelector("[name='analyse']").disabled;`});
 
-    if (logged)
-      await browser.tabs.executeScript({
-        code: `document.getElementById("form3-analyse").click();`
-      });
+      if (loggedIn)
+        await waitAndClick("[name='analyse']");
 
-    await browser.tabs.executeScript({
-      code:
-        "document.getElementById('form3-pgn').value = `" + pgn + "`; document.getElementsByClassName('submit')[0].click();",
-    });
+      if (await waitForElement("[name='pgn']")) {
```

Comment:
Notice how you are building a pyramid of doom here by indenting each level of `if` :smile: This makes the code harder to read since you need to be weary of what code is inside which `if` and if there is code in some `else` (there is none here, but I'm making a general point about such pyramids)

Consider:

1. Inverting `if` conditions and returning early instead of indenting

    ```ts
    if (!(await waitForElement("[name='pgn']")) {
      return;
    }
    
    // the rest
    ```

2. Using rejected promises to stop the code from continuing if there is an unrecoverable failure. See my other comment on line 36 for more details